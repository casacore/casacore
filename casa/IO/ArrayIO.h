//# ArrayIO.h: text output and binary IO for an array of any dimensionality.
//# Copyright (C) 1993,1994,1995,1997,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: casa-feedback@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        charlottesville, VA 22903-2475 USA

#ifndef CASA_ARRAYIO_2_H
#define CASA_ARRAYIO_2_H

//# Includes
#include <vector>
#include <ostream>
#include <regex>

#include <casacore/casa/Arrays/ArrayFwd.h>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

class AipsIO;
class LogIO;
class IPosition;
template<typename T> class Block;
class String;

// <summary>
//    Input/output operators for Arrays.
// </summary>

// <use visibility=export>

// <reviewed reviewer="Paul Shannon" date="1995/02/21" tests="" demos="">
// This header was reviewed and revised with the goal of making it an
// example for those writing global function header files.
// </reviewed>

// <prerequisite>
//   <li> <linkto class=Array>Array</linkto>
//   <li> ostream
//   <li> <linkto class=AipsIO>AipsIO</linkto>
// </prerequisite>

// <etymology>
//  ArrayIO is simply the conventional shorthand for "array input/output".
// </etymology>

// <synopsis> 
// These global functions provide easy input and output of (possibly)
// large and (possibly) multi-dimensional arrays.  Iteration through
// entire arrays is done behind the scenes, with no effort required
// of the client programmer.  
// These functions are global, rather than member functions of the
// Array class, because of the well-known C++ requirement that the first
// argument to an operator function (as it is declared) is the 
// left operand when the function is called.
// </synopsis>


// <example>
// <srcblock>
// IPosition shape (3,10,10,3);
// Array <float> array (shape);
// // ...initialize and manipulate the array...
// cout << "result: " << array;
// </srcblock>
//
// <motivation>
// Effortless input/output is clearly a big win.
// </motivation>
//
// <todo asof="1997/01/15">
// </todo>

// <linkfrom anchor="Array IO" classes="Array Vector Matrix Cube">
//    <here>Array IO</here> -- Input/output operators for Arrays.
// </linkfrom>
//
// <group name="Array IO">

// Write a formatted copy of the array to the LogIO output object. Merely calls
// the ostream operator<< in turn.
template<typename T, typename Alloc>
LogIO &operator<<(LogIO &os, const Array<T> &a);

// Read or write a binary representation of an Array to a file. Very
// useful for saving arrays and restoring them later.
// <br>The putArray function is put in for forwards compatibility
// of images (so new images can be read with old release).
//
// <group>

template<typename T, typename Alloc>
AipsIO &operator<< (AipsIO &, const Array<T> &);

template<typename T, typename Alloc>
void putArray (AipsIO &, const Array<T> &, const char* name);

template<typename T, typename Alloc>
AipsIO &operator>> (AipsIO &, Array<T> &);

// </group>

// </group>

// <summary>
// Global functions to read/write binary arrays from/to a file.
// </summary>

// <use visibility=export>

// <reviewed reviewer="Gareth Hunt" date="95Mar31" tests="" demos="">
// </reviewed>

// <synopsis>
// These global functions provide disk read/write functions for an Array of
// binary numbers.  The write operation is useful, for example, to dump an
// image in binary form to disk so that it can be displayed with an external
// utility such as SAOimage.
// </synopsis>

// <example>
// <srcblock>
//    Matrix<float> picture(256, 256); picture = 0.0;
//    String fileName="picture.data";
//
//    // operations to populate picture
//    // ...
//
//    write_array (picture, fileName);
// </srcblock>
// </example>

// <todo asof="">
//   <li> These functions should eventually be replaced with something
//        more sophisticated.
// </todo>

// <linkfrom anchor="Array binary IO" classes="Array Vector Matrix Cube">
//    <here>Array binary IO</here> -- Simple binary input/output for Arrays.
// </linkfrom>

// <group name=Array binary IO>

// Write the values of an array in binary format into a file with
// the given name.
// The values are stored in local format, thus are not converted
// to a canonical format as
// <linkto class="AipsIO:description">AipsIO</linkto>
// does.
// <note role=warning>
// This function is only suitable for built-in data types.
// </note>
// <group>
template <typename T, typename Alloc>
void write_array (const Array<T>& the_array, const std::string& fileName);

template <typename T, typename Alloc>
inline void write_array (const Array<T>& the_array, const char* fileName)
    { write_array (the_array, std::string(fileName)); }
// </group>

// Read the values of an array in binary format from a file with
// the given name.
// The number of values read is the size of the Array, thus the file
// should at least contain that number of values.
// <note role=warning>
// This function is only suitable for built-in data types.
// </note>
// <group>
template <typename T, typename Alloc>
void read_array (Array<T>& the_array, const std::string& fileName);

template <typename T, typename Alloc>
inline void read_array (Array<T>& the_array, const char* fileName)
    { read_array (the_array, std::string(fileName)); }
// </group>

// </group>

// These two functions read and write a Vector of data.  The input
// may be arranged in any format (i.e. It may be recorded as one value per
// line or it may be recorded with all values on a single line).
// Values must be separated by whitespace.

// <group>
template <typename T, typename Alloc>
void readAsciiVector (Vector<T>& vec, const char* fileName);

template <typename T, typename Alloc>
void writeAsciiVector (const Vector<T>& vec, const char* fileName);
// </group>

// </group>

AipsIO& operator<< (AipsIO& aio, const IPosition& ip);
AipsIO& operator>> (AipsIO& aio, IPosition& ip);
LogIO& operator<< (LogIO& os, const IPosition& ip);

template<typename T, typename Alloc>
Block<T> makeBlock(const Array<T>& array);

template<typename T>
Vector<T> makeVector(const Block<T>& block);

Vector<String> stringToVector (const String& string, char delim = ',');
Vector<String> stringToVector (const String& string, const std::regex& delim);

} //# NAMESPACE CASACORE - END

#include <casacore/casa/IO/ArrayIO.tcc>

#endif
