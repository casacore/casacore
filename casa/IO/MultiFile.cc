//# MultiFile.cc: Class to combine multiple files in a single one
//# Copyright (C) 2014
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: RegularFileIO.h 20551 2009-03-25 00:11:33Z Malte.Marquarding $

//# Includes
#include <casacore/casa/IO/MultiFile.h>
#include <casacore/casa/IO/RegularFileIO.h>
#include <casacore/casa/IO/MemoryIO.h>
#include <casacore/casa/IO/CanonicalIO.h>
#include <casacore/casa/IO/AipsIO.h>
#include <casacore/casa/BasicSL/STLIO.h>
#include <casacore/casa/OS/CanonicalConversion.h>
#include <casacore/casa/Utilities/GenSort.h>
#include <casacore/casa/Utilities/Assert.h>
#include <casacore/casa/Exceptions/Error.h>
#include <unistd.h>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

  MultiFile::MultiFile (const String& name, ByteIO::OpenOption option,
                        Int blockSize)
    : MultiFileBase (name, blockSize)
  {
    itsFD = RegularFileIO::openCreate (itsName, option);
    itsIO.attach (itsFD, itsName);
    if (option == ByteIO::New  ||  option == ByteIO::NewNoReplace) {
      // New file; first block is for administration.
      setNewFile();
      itsNrBlock = 1;
    } else {
      readHeader();
    }
    itsWritable = itsIO.isWritable();
  }

  MultiFile::~MultiFile()
  {
    close();
  }

  void MultiFile::flushFile()
  {
    itsIO.flush();
  }

  void MultiFile::close()
  {
    flush();
    FiledesIO::close (itsFD);
  }

  void MultiFile::reopenRW()
  {
    if (isWritable()) {
      return;
    }
    // First try if the file can be opened as read/write.
    int fd = RegularFileIO::openCreate (itsName, ByteIO::Update);
    // Now close the readonly file and reset fd.
    FiledesIO::close (itsFD);
    itsIO.detach();
    itsFD = fd;
    itsIO.attach (itsFD, itsName);
    itsIO.setWritable();
    itsWritable = True;
  }

  void MultiFile::fsync()
  {
    itsIO.fsync();
  }

  void MultiFile::writeHeader()
  {
    // Write all header info into a memory buffer.
    // This buffer will be written into the file.
    // If needed, spread over multiple blocks.
    MemoryIO mio(itsBlockSize, itsBlockSize);
    CanonicalIO cio(&mio);
    AipsIO aio(&cio);
    itsHdrCounter++;
    cio.write (1, &itsBlockSize);         // reserve space for header size
    cio.write (1, &itsBlockSize);
    cio.write (1, &itsHdrCounter);
    aio.putstart ("MultiFile", 1);
    aio << itsNrBlock << itsInfo << itsFreeBlocks;
    aio.putend();
    Int64 todo = mio.length();
    uChar* buf = const_cast<uChar*>(mio.getBuffer());
    CanonicalConversion::fromLocal (buf, todo);       // header size
    // Write the first part of the buffer at the beginning of the file.
    itsIO.seek (0);
    itsIO.write (itsBlockSize, buf);
    todo -= itsBlockSize;
    if (todo > 0) {
      // The rest is written in another file. If the header info was written
      // at the end of the file, it would be overwritten when extending with
      // possible file corruption if the program or system crashes.
      // By using a separate file, corruption chances are much lower.
      // Even better would be using another name and doing a rename at the end.
      int fd = RegularFileIO::openCreate (itsName + "_hdrext", ByteIO::New);
      FiledesIO iohdr (fd, itsName + "_hdrext");
      iohdr.write (todo, buf+itsBlockSize);
      FiledesIO::close (fd);
    }
  }

  void MultiFile::readHeader (Bool always)
  {
    // Read the first part of the header.
    vector<char> buf(3*sizeof(Int64));
    itsIO.seek (0);
    itsIO.read (buf.size(), &(buf[0]));
    // Extract the required info.
    Int64 headerSize, hdrCounter;
    CanonicalConversion::toLocal (headerSize, &(buf[0]));
    CanonicalConversion::toLocal (itsBlockSize, &(buf[1*sizeof(Int64)]));
    CanonicalConversion::toLocal (hdrCounter, &(buf[2*sizeof(Int64)]));
    // Only if needed, read the rest of the header.
    if (hdrCounter == itsHdrCounter  &&  !always) {
      return;
    }
    itsHdrCounter = hdrCounter;
    Int64 leadSize = 3*sizeof(Int64);
    buf.resize (headerSize);
    if (headerSize > itsBlockSize) {
      itsIO.read (itsBlockSize - leadSize, &(buf[leadSize]));
      int fd = RegularFileIO::openCreate (itsName + "_hdrext", ByteIO::Old);
      FiledesIO iohdr (fd, itsName + "_hdrext");
      iohdr.read (headerSize - itsBlockSize, &(buf[itsBlockSize]));
      FiledesIO::close (fd);
    } else {
      itsIO.read (headerSize - leadSize, &(buf[leadSize]));
    }
    // Read all header info from the memory buffer.
    MemoryIO mio(&(buf[leadSize]), headerSize - leadSize);
    CanonicalIO cio(&mio);
    AipsIO aio(&cio);
    Int version = aio.getstart ("MultiFile");
    AlwaysAssert (version==1, AipsError);
    aio >> itsNrBlock >> itsInfo >> itsFreeBlocks;
    aio.getend();
    // Initialize remaining info fields.
    for (vector<MultiFileInfo>::iterator iter=itsInfo.begin();
         iter!=itsInfo.end(); ++iter) {
      iter->curBlock = -1;
      iter->dirty    = False;
      iter->buffer.resize (itsBlockSize);
    }
  }

  void MultiFile::doAddFile (MultiFileInfo&)
  {}

  void MultiFile::doDeleteFile (MultiFileInfo& info)
  {
    // Add the blocknrs to the free list.
    // Later we can merge them in order and leave out blocks past last block used.
    itsFreeBlocks.reserve (itsFreeBlocks.size() + info.blockNrs.size());
    for (size_t i=0; i<info.blockNrs.size(); ++i) {
      itsFreeBlocks.push_back (info.blockNrs[i]);
    }
    // Sort them in descending order, so free blocks can be taken from the tail.
    genSort (&(itsFreeBlocks[0]), itsFreeBlocks.size(),
             Sort::Descending, Sort::QuickSort);
  }

  void MultiFile::extend (MultiFileInfo& info, Int64 lastblk)
  {
    Int64 curnrb = info.blockNrs.size();
    info.blockNrs.resize (lastblk);
    for (Int64 i=curnrb; i<lastblk; ++i) {
      if (itsFreeBlocks.empty()) {
        info.blockNrs[i] = itsNrBlock;
        itsNrBlock++;
      } else {
        info.blockNrs[i] = itsFreeBlocks[itsFreeBlocks.size() - 1];
        itsFreeBlocks.resize (itsFreeBlocks.size() - 1);
      }
    }
  }

  void MultiFile::readBlock (MultiFileInfo& info, Int64 blknr,
                             void* buffer)
  {
    itsIO.seek (info.blockNrs[blknr] * itsBlockSize);
    itsIO.read (itsBlockSize, buffer);
  }

  void MultiFile::writeBlock (MultiFileInfo& info, Int64 blknr,
                              const void* buffer)
  {
    itsIO.seek  (info.blockNrs[blknr] * itsBlockSize);
    itsIO.write (itsBlockSize, buffer);
  }


} //# NAMESPACE CASACORE - END
