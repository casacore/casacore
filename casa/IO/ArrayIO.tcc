//# ArrayIO.cc: text output and binary IO for an array of any dimensionality.
//# Copyright (C) 1993,1994,1995,1996,1997,1999,2000,2001,2002,2003
//# Associated Universities, Inc. Washington DC, USA.
//# 
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//# 
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//# 
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//# 
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: casa-feedback@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA

#ifndef CASA_ARRAYIO_2_TCC
#define CASA_ARRAYIO_2_TCC

//# Includes
#include <casacore/casa/Arrays/Array.h>
#include <casacore/casa/Arrays/ArrayError.h>
#include <casacore/casa/Arrays/Matrix.h>
#include <casacore/casa/Arrays/Vector.h>
#include <casacore/casa/Arrays/ArrayPosIter.h>
#include <casacore/casa/Arrays/ArrayUtil.h>

#include <casacore/casa/IO/AipsIO.h>
#include <casacore/casa/IO/AipsIOCarray.h>
#include <casacore/casa/IO/ArrayIO.h>

#include <casacore/casa/Logging/LogIO.h>

#include <istream>
#include <fstream>
#include <cassert>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

template<class T>
LogIO &operator<<(LogIO &os, const Array<T> &a)
{
    os.output() << a;
    return os;
}

template<class T>
AipsIO &operator<<(AipsIO &ios, const Array<T> &a)
{
    putArray (ios, a, "Array");
    return ios;
}

template<class T>
void putArray (AipsIO &ios, const Array<T> &a, const Char* name)
{
    if (a.size() * sizeof(T) > 2147483647) {
      throw AipsError("AipsIO putArray too large (exceeds 2**31 bytes)");
    }
    ios.putstart(name, Array<T>::arrayVersion());
    // Write out dimensionality
    ios << uInt(a.ndim());
    // Write out length
    for (size_t i=0; i < a.ndim(); i++) {
      ios << uInt(a.shape()(i));
    }
    // Now write out the data
    bool deleteIt;
    const T *storage =  a.getStorage(deleteIt);
    putAipsIO (ios, a.nelements(), storage);
    a.freeStorage(storage, deleteIt);
    ios.putend();
}

// <thrown>
//     <item> ArrayError
// </thrown>
template<class T>
AipsIO &operator>>(AipsIO &ios, Array<T> &a)
{
    // Makes the argument unique, i.e. existing refs will be lost. At the moment
    // this is relatively inefficient as it makes a temporary copy.
    a.unique();
    
    // On 20-Nov-2000 use of the home-brew rtti was removed.
    // It meant that a name 'Array<int>' is now replaced by 'Array'.
    // In order to recognize those old names, we must do something special.
    String type = ios.getNextType();
    int version;
    if (type.length() > 6  &&  type.index("Array<") == 0) {
      version = ios.getstart(type);
    } else {
      version = ios.getstart("Array");
    }
    int ndim;
    ios >> ndim;
    IPosition shape(ndim);
    // Older versions contain an origin (which we discard).
    if (version < 3) {
	int orig;
	for (int i=0; i < ndim; i++) {
	    ios >> orig;
	}
    }
    uInt v;
    for (int i=0; i < ndim; i++) {
      ios >> v;
      shape(i) = v;
    }
    a.resize(shape);                // hopefully a no-op if unchanged

    // Now read in the data.

    bool deleteIt;
    T *storage = a.getStorage(deleteIt);

    uInt nwritten;
    ios >> nwritten;
    if (nwritten != a.nelements())
	throw(ArrayError("AipsIO &operator>>(AipsIO, Array<T>"
			  " - nelements() differs from number in file"));
    getAipsIO (ios, nwritten, storage);
    a.putStorage(storage, deleteIt);
    ios.getend();
    return ios;
}

inline LogIO& operator<< (LogIO& os, const IPosition& ip)
{
    os.output() << ip;
    return os;
}

template<typename T>
Block<T> makeBlock(const Array<T>& array)
{
	Block<T> block(array.nelements());
	if(array.contiguousStorage())
		std::copy(array.cbegin(), array.cend(), block.storage());
	else
		std::copy(array.begin(), array.end(), block.storage());
	return block;
}

template<typename T>
Vector<T> makeVector(const Block<T>& block)
{
	return Vector<T>(block.begin(), block.end());
}

inline Vector<String> stringToVector (const String& string, char delim)
{
	Vector<std::string> vec = strToVector(string, delim);
	return Vector<String>(vec.begin(), vec.end());
}

inline Vector<String> stringToVector (const String& string, const std::regex& delim)
{
	Vector<std::string> vec = strToVector(string, delim);
	return Vector<String>(vec.begin(), vec.end());
}

} //# NAMESPACE CASACORE - END


#endif
