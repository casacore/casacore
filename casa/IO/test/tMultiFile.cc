//# tMultiFile.cc: Test program for class MultiFile
//# Copyright (C) 2014
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: RegularFileIO.h 20551 2009-03-25 00:11:33Z Malte.Marquarding $

//# Includes
#include <casa/IO/MultiFile.h>
#include <casa/Utilities/Assert.h>
#include <casa/Arrays/Vector.h>
#include <casa/Arrays/ArrayMath.h>
#include <casa/Arrays/ArrayLogical.h>
#include <casa/Arrays/ArrayIO.h>
#include <casa/BasicSL/STLIO.h>
#include <iostream>
#include <stdexcept>

using namespace casa;
using namespace std;

void showMultiFile (MultiFile& mfile)
{
  cout << mfile.fileName() << ' ' << mfile.blockSize() << ' '
       << mfile.nfile() << ' ' << mfile.size() << endl;
}

void makeFile()
{
  MultiFile mfile("tMultiFile_tmp.dat", ByteIO::New, 1024);
  AlwaysAssertExit (mfile.isWritable());
  showMultiFile(mfile);
}

void readFile()
{
  MultiFile mfile("tMultiFile_tmp.dat", ByteIO::Old);
  AlwaysAssertExit (! mfile.isWritable());
  showMultiFile(mfile);
  for (Int i=0; i<mfile.nfile(); ++i) {
    String nm = "file" + String::toString(i);
    cout << nm << ' ' << mfile.fileId(nm) << endl;
  }
}

void addFiles()
{
  MultiFile mfile("tMultiFile_tmp.dat", ByteIO::Update);
  AlwaysAssertExit (mfile.isWritable());
  Int fid0 = mfile.add ("file0");
  Int fid1 = mfile.add ("file1");
  Int fid2 = mfile.add ("file2");
  AlwaysAssertExit (mfile.nfile()==3 && fid0==0 && fid1==1 && fid2==2);
  showMultiFile(mfile);
}

void writeFiles1()
{
  MultiFile mfile("tMultiFile_tmp.dat", ByteIO::Update);
  Vector<Int64> buf(128);
  indgen(buf);
  mfile.write (0, buf.data(), 1024, 0);
  buf += Int64(128);
  mfile.write (2, buf.data(), 1024, 0);
  buf += Int64(128);
  mfile.write (0, buf.data(), 1024, 1024);
  buf += Int64(128);
  mfile.write (0, buf.data(), 1024, 2048);
  buf += Int64(128);
  mfile.write (1, buf.data(), 1024, 1024);
  buf += Int64(128);
  mfile.write (2, buf.data(), 1024, 1024);
  cout << mfile.info() << endl;
}

void checkFiles1()
{
  MultiFile mfile("tMultiFile_tmp.dat", ByteIO::Old);
  Vector<Int64> buf1(128), buf(128),buff(3*128);
  indgen(buf1);
  mfile.read (0, buf.data(), 1024, 0);
  AlwaysAssertExit (allEQ(buf, buf1));
  buf1 += Int64(128);
  mfile.read (2, buf.data(), 1024, 0);
  AlwaysAssertExit (allEQ(buf, buf1));
  buf1 += Int64(128);
  mfile.read (0, buf.data(), 1024, 1024);
  AlwaysAssertExit (allEQ(buf, buf1));
  buf1 += Int64(128);
  mfile.read (0, buf.data(), 1024, 2048);
  AlwaysAssertExit (allEQ(buf, buf1));
  buf1 += Int64(128);
  mfile.read (1, buf.data(), 1024, 1024);
  AlwaysAssertExit (allEQ(buf, buf1));
  buf1 += Int64(128);
  mfile.read (2, buf.data(), 1024, 1024);
  AlwaysAssertExit (allEQ(buf, buf1));
  // Check a single read.
  indgen(buf1);
  mfile.read (0, buff.data(), 3072, 0);
  AlwaysAssertExit (allEQ(buff(Slice(0,128)), buf1));
  AlwaysAssertExit (allEQ(buff(Slice(128,128)), buf1+Int64(256)));
  AlwaysAssertExit (allEQ(buff(Slice(256,128)), buf1+Int64(384)));
  mfile.read (0, buff.data(), 3072-24, 8);
  AlwaysAssertExit (allEQ(buff(Slice(0,127)), buf1(Slice(1,127))));
  AlwaysAssertExit (allEQ(buff(Slice(127,128)), buf1+Int64(256)));
  AlwaysAssertExit (allEQ(buff(Slice(255,126)), buf1(Slice(0,126))+Int64(384)));
  AlwaysAssertExit (buff[380]==509 && buff[381]==509);   // check not overwritten
}

int main()
{
  try {
    makeFile();
    readFile();
    addFiles();
    readFile();
    writeFiles1();
    readFile();
    checkFiles1();
  } catch (std::exception& x) {
    cout << "Unexpected exception: " << x.what() << endl;
    return 1;
  }
  return 0;
}
