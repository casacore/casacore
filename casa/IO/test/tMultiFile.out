MultiFile test with blockSize=1024, useCRC=0, useODirect=0
tMultiFile_tmp.dat: blocksize=1024  nfile=0  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tMultiFile_tmp.dat: blocksize=1024  nfile=0  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tMultiFile_tmp.dat: blocksize=1024  nfile=3  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tMultiFile_tmp.dat: blocksize=1024  nfile=3  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
file0 0
file1 1
file2 2
[file0 [1,3,4]
  fsize=3072 cur=-1 nest=0 dirty=0
,file1 [5,6]
  fsize=2048 cur=-1 nest=0 dirty=0
,file2 [2,7]
  fsize=2048 cur=-1 nest=0 dirty=0
]
tMultiFile_tmp.dat: blocksize=1024  nfile=3  nblock=8  nCRC=0
  ncont=0,0  cont=0 []  free=[]
file0 0
file1 1
file2 2
test deleteFile
tMultiFile_tmp.dat: blocksize=1024  nfile=2  nblock=8  nCRC=0
  ncont=0,0  cont=0 []  free=[6,5]
file0 0
file1 -1
file2 2
[file0 [1,3,4]
  fsize=3072 cur=0 nest=0 dirty=1
, []
  fsize=0 cur=-1 nest=0 dirty=0
,file2 [2,7,5]
  fsize=2064 cur=2 nest=0 dirty=1
]
tMultiFile_tmp.dat: blocksize=1024  nfile=2  nblock=8  nCRC=0
  ncont=0,0  cont=0 []  free=[6]
file0 0
file1 -1
file2 2

MultiFile test with blockSize=128, useCRC=0, useODirect=0
tMultiFile_tmp.dat: blocksize=128  nfile=0  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tMultiFile_tmp.dat: blocksize=128  nfile=0  nblock=2  nCRC=0
  ncont=0,1  cont=1 [1]  free=[]
tMultiFile_tmp.dat: blocksize=128  nfile=3  nblock=2  nCRC=0
  ncont=0,1  cont=1 [1]  free=[]
tMultiFile_tmp.dat: blocksize=128  nfile=3  nblock=3  nCRC=0
  ncont=1,1  cont=0 [2]  free=[]
file0 0
file1 1
file2 2
[file0 [3,4,5,6,7,8,9,10,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34]
  fsize=3072 cur=-1 nest=0 dirty=0
,file1 [35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50]
  fsize=2048 cur=-1 nest=0 dirty=0
,file2 [11,12,13,14,15,16,17,18,51,52,53,54,55,56,57,58]
  fsize=2048 cur=-1 nest=0 dirty=0
]
tMultiFile_tmp.dat: blocksize=128  nfile=3  nblock=60  nCRC=0
  ncont=1,2  cont=1 [1,59]  free=[]
file0 0
file1 1
file2 2
test deleteFile
tMultiFile_tmp.dat: blocksize=128  nfile=2  nblock=62  nCRC=0
  ncont=3,2  cont=0 [2,60,61]  free=[50,49,48,47,46,45,44,43,42,41,40,39,38,37,36,35]
file0 0
file1 -1
file2 2
[file0 [3,4,5,6,7,8,9,10,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34]
  fsize=3072 cur=0 nest=0 dirty=1
, []
  fsize=0 cur=-1 nest=0 dirty=0
,file2 [11,12,13,14,15,16,17,18,51,52,53,54,55,56,57,58,35]
  fsize=2064 cur=16 nest=0 dirty=1
]
tMultiFile_tmp.dat: blocksize=128  nfile=2  nblock=63  nCRC=0
  ncont=3,3  cont=1 [1,59,62]  free=[50,49,48,47,46,45,44,43,42,41,40,39,38,37,36]
file0 0
file1 -1
file2 2

MultiFile test with blockSize=4096, useCRC=1, useODirect=1
tMultiFile_tmp.dat: blocksize=4096  nfile=0  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tMultiFile_tmp.dat: blocksize=4096  nfile=0  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tMultiFile_tmp.dat: blocksize=4096  nfile=3  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tMultiFile_tmp.dat: blocksize=4096  nfile=3  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
file0 0
file1 1
file2 2
[file0 [1]
  fsize=3072 cur=0 nest=0 dirty=1
,file1 [3]
  fsize=2048 cur=0 nest=0 dirty=1
,file2 [2]
  fsize=2048 cur=0 nest=0 dirty=1
]
tMultiFile_tmp.dat: blocksize=4096  nfile=3  nblock=4  nCRC=4
  ncont=0,0  cont=0 []  free=[]
file0 0
file1 1
file2 2
test deleteFile
tMultiFile_tmp.dat: blocksize=4096  nfile=2  nblock=3  nCRC=3
  ncont=0,0  cont=0 []  free=[]
file0 0
file1 -1
file2 2
[file0 [1]
  fsize=3072 cur=0 nest=0 dirty=1
, []
  fsize=0 cur=-1 nest=0 dirty=0
,file2 [2]
  fsize=2064 cur=0 nest=0 dirty=1
]
tMultiFile_tmp.dat: blocksize=4096  nfile=2  nblock=3  nCRC=3
  ncont=0,0  cont=0 []  free=[]
file0 0
file1 -1
file2 2

Test a version 1 MultiFile ...
tMultiFile.in: blocksize=128  nfile=2  nblock=57  nCRC=0
  ncont=0,0  cont=0 []  free=[48,47,46,45,44,43,42,41,40,39,38,37,36,35,34]
file0 0
file1 -1
file2 2

Test nested with block sizes 512 and 0
tMultiFile_tmp.nest: blocksize=512  nfile=2  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tnested: blocksize=512  nfile=1  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tMultiFile_tmp.nest: blocksize=512  nfile=2  nblock=2  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tnested: blocksize=512  nfile=1  nblock=1  nCRC=0
  ncont=0,0  cont=0 []  free=[]

Test truncation of MultiFile (also nested) ...
tMultiFile_tmp.dat: blocksize=256  nfile=1  nblock=11  nCRC=0
  ncont=0,0  cont=0 []  free=[]
tMultiFile_tmp.dat: blocksize=256  nfile=1  nblock=5  nCRC=0
  ncont=0,0  cont=0 []  free=[]
4
>>>
pack            0 real           0 user           0 system
unpack          0 real           0 user           0 system
pack         0.05 real        0.04 user           0 system
unpack       0.05 real        0.04 user        0.01 system
move1        0.03 real        0.02 user           0 system
move2        0.05 real        0.05 user           0 system
move2        0.05 real        0.05 user           0 system
move3        0.12 real        0.11 user           0 system
exact        0.02 real           0 user        0.02 system
exact        0.02 real           0 user           0 system
double       0.02 real           0 user        0.01 system
double       0.01 real        0.01 user        0.01 system
partly       0.01 real           0 user        0.01 system
partly       0.04 real           0 user        0.02 system
<<<
tMultiFile ended OK
